File: backend/src/main/java/com/deviceagent/memory/ContextBuilder.java
package com.deviceagent.memory;

import com.deviceagent.domain.StateSnapshot;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Minimal context injection for each model decision (docs/03 §14.2).
 */
@Component
public class ContextBuilder {
    private final MemoryRetriever retriever;

    public ContextBuilder(MemoryRetriever retriever) {
        this.retriever = retriever;
    }

    public BuiltContext build(
            String sessionId,
            String userText,
            List<Map<String, Object>> goals,
            List<Map<String, Object>> constraints,
            List<Map<String, Object>> criteria,
            StateSnapshot observation,
            List<Map<String, Object>> priorActions
    ) {
        List<MemoryEntry> memories = retriever.retrieve(sessionId, userText, goals);
        List<Map<String, Object>> hints = new ArrayList<>();
        for (MemoryEntry m : memories) {
            hints.add(m.toHint());
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("goals", goals == null ? List.of() : goals);
        payload.put("constraints", constraints == null ? List.of() : constraints);
        payload.put("criteria", criteria == null ? List.of() : criteria);
        payload.put("observation", observation == null ? Map.of() : observation.getState());
        payload.put("prior_actions_summary", summarizeActions(priorActions));
        payload.put("long_term_memory", hints);
        payload.put("token_budget_note", "minimal_context_v1");
        int approxTokens = estimateTokens(payload);
        payload.put("approx_tokens", approxTokens);
        return new BuiltContext(payload, memories, hints, approxTokens);
    }

    public BuiltContext buildForCompile(String sessionId, String userText, StateSnapshot observation) {
        return build(sessionId, userText, List.of(), List.of(), List.of(), observation, List.of());
    }

    private List<Map<String, Object>> summarizeActions(List<Map<String, Object>> priorActions) {
        if (priorActions == null) return List.of();
        List<Map<String, Object>> out = new ArrayList<>();
        int from = Math.max(0, priorActions.size() - 6);
        for (int i = from; i < priorActions.size(); i++) {
            Map<String, Object> a = priorActions.get(i);
            Map<String, Object> s = new LinkedHashMap<>();
            s.put("capability_id", a.get("capability_id"));
            s.put("execution_status", a.get("execution_status"));
            s.put("params", a.get("params"));
            out.add(s);
        }
        return out;
    }

    private int estimateTokens(Map<String, Object> payload) {
        return Math.max(1, String.valueOf(payload).length() / 3);
    }

    public record BuiltContext(
            Map<String, Object> payload,
            List<MemoryEntry> memories,
            List<Map<String, Object>> hints,
            int approxTokens
    ) {}
}

File: backend-py/src/terminal_agent/memory/context_builder.py
"""Minimal context injection for each model decision (docs/03 §14.2)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from terminal_agent.contracts import StateSnapshot
from terminal_agent.memory.entry import MemoryEntry
from terminal_agent.memory.retriever import MemoryRetriever


@dataclass(frozen=True)
class BuiltContext:
    payload: dict[str, Any]
    memories: list[MemoryEntry]
    hints: list[dict[str, Any]]
    approx_tokens: int


class ContextBuilder:
    def __init__(self, retriever: MemoryRetriever) -> None:
        self.retriever = retriever

    def build(
        self,
        session_id: str | None,
        user_text: str | None,
        goals: list[dict[str, Any]] | None,
        constraints: list[dict[str, Any]] | None,
        criteria: list[dict[str, Any]] | None,
        observation: StateSnapshot | None,
        prior_actions: list[dict[str, Any]] | None,
        tenant_id: str | None = None,
    ) -> BuiltContext:
        memories = self.retriever.retrieve(session_id, user_text, goals, tenant_id)
        hints = [m.to_hint() for m in memories]
        payload: dict[str, Any] = {
            "goals": goals or [],
            "constraints": constraints or [],
            "criteria": criteria or [],
            "observation": observation.state if observation is not None else {},
            "prior_actions_summary": self._summarize_actions(prior_actions),
            "long_term_memory": hints,
            "token_budget_note": "minimal_context_v1",
        }
        approx_tokens = self._estimate_tokens(payload)
        payload["approx_tokens"] = approx_tokens
        return BuiltContext(payload, memories, hints, approx_tokens)

    def build_for_compile(
        self,
        session_id: str | None,
        user_text: str | None,
        observation: StateSnapshot | None,
        tenant_id: str | None = None,
    ) -> BuiltContext:
        return self.build(session_id, user_text, [], [], [], observation, [], tenant_id)

    @staticmethod
    def _summarize_actions(prior_actions: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
        if not prior_actions:
            return []
        start = max(0, len(prior_actions) - 6)
        out: list[dict[str, Any]] = []
        for action in prior_actions[start:]:
            out.append(
                {
                    "capability_id": action.get("capability_id"),
                    "execution_status": action.get("execution_status"),
                    "params": action.get("params"),
                }
            )
        return out

    @staticmethod
    def _estimate_tokens(payload: dict[str, Any]) -> int:
        return max(1, len(str(payload)) // 3)
